timesketch/lib/analyzers/geoip.py

```diff
@@ -0,0 +1,309 @@
+            raise RuntimeError('MaxMind host not set.')
+        super().__init__(self._account_id, self._license_key, host=self._host)
+
+    def ip2geo(self, ip_address) -> Union[Tuple[str, str, str, str], None]:
+        """Perform a IP to geolocation lookup.
+
+        Args:
+            ip_address - the IPv4 or IPv6 address to geolocate
+
+        Returns:
+            Either:
+            A tuple comprising of the following in order
+            - iso_code (str) - the ISO 31661 alpha-2 code of the country
+            - latitude (str) - the north-south coordinate
+            - longitude (str) - the east-west coordinate
+            - country_name (str) - the full country name
+            - city (str) - the city name that approximates the location
+            Or None:
+            - when the IP address does not have a resolvable location
+        """
+        try:
+            response = self.city(ip_address)
+        except geoip2.errors.AddressNotFoundError as exception:
+            logging.debug('IP address {0} not found.'.format(
+                ip_address))
+            return None
+        except Exception as error:
+            logging.error('Error while geolocating {0} - {1}'.format(
+                ip_address, error))
+            return None
+
+        latitude = response.location.latitude
+        longitude = response.location.longitude
+        country_name = response.country.name
+        iso_code = response.country.iso_code
+        city = response.city.name
+
+        return (iso_code, latitude, longitude, country_name, city)
+
+
+class BaseGeoIpAnalyzer(interface.BaseAnalyzer):
+    """Sketch analyzer for geolocating IP addresses.
+
+    Concrete plugin implementations should define the following attributes
+    NAME, DISPLAY_NAME, DESCRIPTION, GEOIP_CLIENT.
+    """
+
+    NAME = ""
+    DISPLAY_NAME = ""
+    DESCRIPTION = ""
+
+    GEOIP_CLIENT = None
+
+    DEPENDENCIES = frozenset(['feature_extraction'])
+    IP_FIELDS = ['ip', 'host_ip', 'src_ip', 'dst_ip', 'source_ip', 'dest_ip',
+        'ip_address', 'client_ip', 'address', 'saddr', 'daddr',
+        'requestMetadata_callerIp', 'a_answer']
+
+
+    def __init__(self, index_name, sketch_id, timeline_id=None):
+        """Initialize The Sketch Analyzer.
+        Args:
+            index_name: Elasticsearch index name
+            sketch_id: Sketch ID
+        """
+        self.index_name = index_name
+        super().__init__(index_name, sketch_id, timeline_id=timeline_id)
+
+
+    def _validate_ip(self, ip_address):
+        """Return true if ip_address is in a valid format and is non-private."""
+        try:
+            ip = ipaddress.ip_address(ip_address)
+            return ip.is_global
+        except ValueError as exception:
+            return False
+
+    def run(self):
+        """Entry point for the analyzer.
+
+        Returns:
+            String with summary of the analyzer result
+        """
+        query = ' OR '.join(
```

Comment:
You can check if an attribute exists by searching for `_exists_:field_name`. I would opt to use that. As an example:

```Example
        _exists_:(ip OR src_ip OR dst_ip)
```

Reply:
Implemented as suggested.